java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/EncoderUtils.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import org.apache.arrow.util.DataSizeRoundingUtil;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.ValueVector;
+
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * Utilities for dictionnary encoding/decoding.
+ */
+public final class EncoderUtils {
+
+  /**
+   * Decodes an input vector. The algorithm takes O(n) time,
+   * where n is the length of the input vector.
+   * @param input the input vector.
+   * @param output the output vector.
+   * @param dictionary the dictionary.
+   * @param encodeNull a flag indicating if null should be decoded.
+   */
+  public static <E extends BaseIntVector, D extends ValueVector> void decode(
+          E input, D output, D dictionary, boolean encodeNull) {
+    for (int i = 0; i < input.getValueCount(); i++) {
+      if (!encodeNull && input.isNull(i)) {
+        continue;
+      }
+
+      int index = (int) input.getValueAsLong(i);
+      output.copyFromSafe(index, i, dictionary);
+    }
+    
+    // if encode null is disabled, the validity bits for input/output
+    // vectors are identical, so we can improve performance by copying
+    // the buffer directly.
+    if (!encodeNull) {
+      // make sure the output vector is large enough
```

Comment:
does this replicate code in setValue count?

Reply:
Good point. Thank you.
setValueCount will do the necessary reallocate.